File: backend-spring-boot/src/main/java/no/vicx/backend/error/GlobalExceptionHandler.java
package no.vicx.backend.error;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.ConstraintViolationException;
import jakarta.validation.Path;
import org.springframework.http.HttpStatus;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@RestControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    ApiError handleValidationException(MethodArgumentNotValidException exception, HttpServletRequest request) {
        return new ApiError(
                HttpStatus.BAD_REQUEST.value(),
                "validation error",
                request.getServletPath(),
                exception.getBindingResult().getFieldErrors().stream()
                        .collect(Collectors.toMap(
                                FieldError::getField,
                                fieldError -> Optional.ofNullable(fieldError.getDefaultMessage()).orElse("Unknown error"),
                                (firstMessage, secondMessage) -> firstMessage)));
    }

    @ExceptionHandler(ConstraintViolationException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    ApiError handleConstraintViolationException(ConstraintViolationException exception, HttpServletRequest request) {
        return new ApiError(
                HttpStatus.BAD_REQUEST.value(),
                "validation error",
                request.getServletPath(),
                exception.getConstraintViolations().stream()
                        .collect(Collectors.toMap(
                                violation -> StreamSupport.stream(violation.getPropertyPath().spliterator(), false)
                                        .reduce((first, second) -> second)
                                        .map(Path.Node::getName)
                                        .orElse("unknown"),
                                ConstraintViolation::getMessage,
                                (firstMessage, secondMessage) -> firstMessage))
        );
    }
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/error/GlobalExceptionHandler.kt
package no.vicx.backend.error

import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleValidationException(
        exception: MethodArgumentNotValidException,
        request: HttpServletRequest,
    ) = ApiError(
        status = HttpStatus.BAD_REQUEST.value(),
        message = VALIDATION_ERROR_MSG,
        url = request.servletPath,
        validationErrors =
            exception.bindingResult.fieldErrors
                .associate { it.field to (it.defaultMessage ?: "Unknown error") },
    )

    @ExceptionHandler(ConstraintViolationException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleConstraintViolationException(
        exception: ConstraintViolationException,
        request: HttpServletRequest,
    ) = ApiError(
        status = HttpStatus.BAD_REQUEST.value(),
        message = VALIDATION_ERROR_MSG,
        url = request.servletPath,
        validationErrors =
            exception.constraintViolations
                .associate { (it.propertyPath.last().name ?: FALLBACK_PROPERTY_PATH) to it.message },
    )

    companion object {
        const val VALIDATION_ERROR_MSG = "validation error"
        const val FALLBACK_PROPERTY_PATH = "ROOT"
    }
}
